File: server/agent/langgraph/agents/ResearchAgent.ts
import { z } from "zod";
import OpenAI from "openai";
import { BaseAgent, BaseAgentConfig, AgentTask, AgentResult, AgentCapability } from "./types";

const xaiClient = new OpenAI({
  baseURL: "https://api.x.ai/v1",
  apiKey: process.env.XAI_API_KEY,
});

const DEFAULT_MODEL = "grok-4-1-fast-non-reasoning";

export class ResearchAgent extends BaseAgent {
  constructor() {
    const config: BaseAgentConfig = {
      name: "ResearchAgent",
      description: "Specialized agent for web research, information gathering, fact-checking, and knowledge synthesis. Expert at finding and analyzing information from multiple sources.",
      model: DEFAULT_MODEL,
      temperature: 0.2,
      maxTokens: 8192,
      systemPrompt: `You are the ResearchAgent - an expert researcher and information analyst.

Your capabilities:
1. Web Search: Find relevant information from the internet
2. Deep Research: Conduct thorough multi-source research
3. Fact Checking: Verify claims and validate information
4. Source Analysis: Evaluate credibility and relevance of sources
5. Knowledge Synthesis: Combine information into coherent insights
6. Citation Management: Track and format references properly

Research methodology:
- Start with broad searches, then narrow down
- Cross-reference multiple sources
- Identify primary vs secondary sources
- Note publication dates and author credentials
- Flag any conflicting information
- Provide confidence levels for findings

Output format:
- Executive summary
- Key findings with sources
- Detailed analysis
- Confidence assessment
- Recommendations for further research`,
      tools: ["search_web", "research_deep", "fetch_url", "browser_extract"],
      timeout: 180000,
      maxIterations: 20,
    };
    super(config);
  }

  async execute(task: AgentTask): Promise<AgentResult> {
    const startTime = Date.now();
    this.updateState({ status: "running", currentTask: task.description, startedAt: new Date().toISOString() });

    try {
      const researchType = this.determineResearchType(task);
      let result: any;

      switch (researchType) {
        case "quick_search":
          result = await this.quickSearch(task);
          break;
        case "deep_research":
          result = await this.deepResearch(task);
          break;
        case "fact_check":
          result = await this.factCheck(task);
          break;
        case "competitive_analysis":
          result = await this.competitiveAnalysis(task);
          break;
        default:
          result = await this.generalResearch(task);
      }

      this.updateState({ status: "completed", progress: 100, completedAt: new Date().toISOString() });

      return {
        taskId: task.id,
        agentId: this.state.id,
        success: true,
        output: result,
        duration: Date.now() - startTime,
      };
    } catch (error: any) {
      this.updateState({ status: "failed", error: error.message });
      return {
        taskId: task.id,
        agentId: this.state.id,
        success: false,
        error: error.message,
        duration: Date.now() - startTime,
      };
    }
  }

  private determineResearchType(task: AgentTask): string {
    const description = task.description.toLowerCase();
    if (description.includes("fact check") || description.includes("verify")) return "fact_check";
    if (description.includes("competitor") || description.includes("market")) return "competitive_analysis";
    if (description.includes("deep") || description.includes("comprehensive")) return "deep_research";
    if (description.includes("quick") || description.includes("simple")) return "quick_search";
    return "general";
  }

  private async quickSearch(task: AgentTask): Promise<any> {
    const response = await xaiClient.chat.completions.create({
      model: this.config.model,
      messages: [
        { role: "system", content: this.config.systemPrompt },
        {
          role: "user",
          content: `Quick search for: ${task.description}
Input: ${JSON.stringify(task.input)}

Provide a concise answer with key facts and sources.`,
        },
      ],
      temperature: 0.2,
    });

    return {
      type: "quick_search",
      query: task.description,
      result: response.choices[0].message.content,
      confidence: 0.85,
      timestamp: new Date().toISOString(),
    };
  }

  private async deepResearch(task: AgentTask): Promise<any> {
    this.updateState({ progress: 10 });

    const planResponse = await xaiClient.chat.completions.create({
      model: this.config.model,
      messages: [
        { role: "system", content: this.config.systemPrompt },
        {
          role: "user",
          content: `Create a research plan for: ${task.description}

Return JSON:
{
  "researchQuestions": ["key questions to answer"],
  "searchQueries": ["specific search queries"],
  "sourceTypes": ["types of sources to consult"],
  "methodology": "research approach"
}`,
        },
      ],
      temperature: 0.2,
    });

    this.updateState({ progress: 30 });

    const planContent = planResponse.choices[0].message.content || "{}";
    const planMatch = planContent.match(/\{[\s\S]*\}/);
    const plan = planMatch ? JSON.parse(planMatch[0]) : { searchQueries: [task.description] };

    const findings: any[] = [];
    for (let i = 0; i < (plan.searchQueries?.length || 1); i++) {
      const query = plan.searchQueries?.[i] || task.description;
      
      const searchResponse = await xaiClient.chat.completions.create({
        model: this.config.model,
        messages: [
          { role: "system", content: "You are researching a specific query. Provide detailed findings with sources." },
          { role: "user", content: `Research: ${query}` },
        ],
        temperature: 0.2,
      });

      findings.push({
        query,
        findings: searchResponse.choices[0].message.content,
      });

      this.updateState({ progress: 30 + Math.round((i + 1) / plan.searchQueries.length * 40) });
    }

    this.updateState({ progress: 80 });

    const synthesisResponse = await xaiClient.chat.completions.create({
      model: this.config.model,
      messages: [
        { role: "system", content: this.config.systemPrompt },
        {
          role: "user",
          content: `Synthesize these research findings into a comprehensive report:

Research Topic: ${task.description}
Findings: ${JSON.stringify(findings, null, 2)}

Provide:
1. Executive Summary
2. Key Findings
3. Detailed Analysis
4. Confidence Assessment
5. Recommendations`,
        },
      ],
      temperature: 0.2,
    });

    return {
      type: "deep_research",
      topic: task.description,
      plan,
      findings,
      synthesis: synthesisResponse.choices[0].message.content,
      confidence: 0.9,
      timestamp: new Date().toISOString(),
    };
  }

  private async factCheck(task: AgentTask): Promise<any> {
    const response = await xaiClient.chat.completions.create({
      model: this.config.model,
      messages: [
        { role: "system", content: `${this.config.systemPrompt}

For fact-checking, evaluate:
1. Source credibility
2. Evidence quality
3. Consistency with other sources
4. Potential biases
5. Date and context relevance` },
        {
          role: "user",
          content: `Fact check this claim: ${task.description}
Additional context: ${JSON.stringify(task.input)}

Return JSON:
{
  "claim": "the claim being checked",
  "verdict": "true|false|partially_true|unverifiable",
  "confidence": 0.0-1.0,
  "evidence": [{"source": "", "finding": "", "supports": boolean}],
  "analysis": "detailed analysis",
  "caveats": ["any important caveats"]
}`,
        },
      ],
      temperature: 0.1,
    });

    const content = response.choices[0].message.content || "{}";
    const jsonMatch = content.match(/\{[\s\S]*\}/);
    
    return {
      type: "fact_check",
      result: jsonMatch ? JSON.parse(jsonMatch[0]) : { analysis: content },
      timestamp: new Date().toISOString(),
    };
  }

  private async competitiveAnalysis(task: AgentTask): Promise<any> {
    const response = await xaiClient.chat.completions.create({
      model: this.config.model,
      messages: [
        { role: "system", content: this.config.systemPrompt },
        {
          role: "user",
          content: `Conduct competitive analysis for: ${task.description}
Context: ${JSON.stringify(task.input)}

Analyze:
1. Key competitors
2. Market positioning
3. Strengths and weaknesses
4. Opportunities and threats
5. Strategic recommendations`,
        },
      ],
      temperature: 0.2,
    });

    return {
      type: "competitive_analysis",
      subject: task.description,
      analysis: response.choices[0].message.content,
      timestamp: new Date().toISOString(),
    };
  }

  private async generalResearch(task: AgentTask): Promise<any> {
    const response = await xaiClient.chat.completions.create({
      model: this.config.model,
      messages: [
        { role: "system", content: this.config.systemPrompt },
        {
          role: "user",
          content: `Research request: ${task.description}
Input: ${JSON.stringify(task.input)}

Provide comprehensive research findings.`,
        },
      ],
      temperature: 0.2,
    });

    return {
      type: "general_research",
      query: task.description,
      result: response.choices[0].message.content,
      timestamp: new Date().toISOString(),
    };
  }

  getCapabilities(): AgentCapability[] {
    return [
      {
        name: "web_search",
        description: "Search the web for information",
        inputSchema: z.object({ query: z.string(), maxResults: z.number().optional() }),
        outputSchema: z.object({ results: z.array(z.any()), summary: z.string() }),
      },
      {
        name: "deep_research",
        description: "Conduct comprehensive multi-source research",
        inputSchema: z.object({ topic: z.string(), depth: z.enum(["shallow", "medium", "deep"]).optional() }),
        outputSchema: z.object({ findings: z.array(z.any()), synthesis: z.string() }),
      },
      {
        name: "fact_check",
        description: "Verify claims and validate information",
        inputSchema: z.object({ claim: z.string(), context: z.string().optional() }),
        outputSchema: z.object({ verdict: z.string(), confidence: z.number(), evidence: z.array(z.any()) }),
      },
    ];
  }
}

export const researchAgent = new ResearchAgent();

File: python_agent_tools/src/agents/research_agent.py
"""Research Agent - Web search, information gathering, and data synthesis."""

from typing import Any, Dict, List, Optional
from .base_agent import BaseAgent, AgentConfig, AgentResult, AgentState
import structlog


class ResearchAgentConfig(AgentConfig):
    """Configuration for the Research Agent."""
    max_search_results: int = 10
    search_depth: int = 2
    enable_synthesis: bool = True
    trusted_sources: List[str] = []


class ResearchAgent(BaseAgent):
    """Agent specialized in web search, information gathering, and data synthesis."""
    
    name = "research"
    
    def __init__(
        self,
        config: Optional[ResearchAgentConfig] = None,
        tools: Optional[List] = None,
        memory = None,
    ):
        super().__init__(tools=tools, memory=memory)
        self.config = config or ResearchAgentConfig(name="research")
        self._search_cache: Dict[str, Any] = {}
    
    @property
    def description(self) -> str:
        return "Performs web searches, gathers information from multiple sources, and synthesizes findings"
    
    @property
    def category(self) -> str:
        return "research"
    
    @property
    def tools_used(self) -> List[str]:
        return ["search_web", "api_call", "embeddings", "reason"]
    
    def get_system_prompt(self) -> str:
        return """You are the Research Agent, specialized in information gathering and synthesis.
Your role is to:
1. Perform comprehensive web searches on given topics
2. Gather information from multiple authoritative sources
3. Verify facts by cross-referencing sources
4. Synthesize findings into clear, structured summaries
5. Cite sources and provide confidence levels
6. Identify knowledge gaps and suggest follow-up research

Research methodology:
- Start with broad searches, then narrow down
- Prioritize recent and authoritative sources
- Cross-reference claims across multiple sources
- Extract key facts, statistics, and quotes
- Organize findings by relevance and reliability

Output format:
- Provide executive summary first
- Include detailed findings with citations
- Note any contradictions or uncertainties
- Suggest areas for further investigation"""
    
    async def search(self, query: str, max_results: Optional[int] = None) -> List[Dict[str, Any]]:
        """Perform a web search."""
        max_results = max_results or self.config.max_search_results
        
        if query in self._search_cache:
            return self._search_cache[query]
        
        result = await self.execute_tool("search_web", {
            "query": query,
            "max_results": max_results
        })
        
        if result.success and result.data is not None:
            self._search_cache[query] = result.data
            return result.data
        return []
    
    async def synthesize(self, findings: List[Dict[str, Any]]) -> str:
        """Synthesize findings into a coherent summary."""
        if not self.config.enable_synthesis:
            return str(findings)
        
        result = await self.execute_tool("reason", {
            "task": "Synthesize the following research findings into a coherent summary",
            "context": {"findings": findings}
        })
        
        if result.success and result.data is not None and isinstance(result.data, dict):
            return result.data.get("synthesis", str(findings))
        return str(findings)
    
    async def run(self, task: str, context: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Execute the research agent's main loop."""
        self.state = AgentState.EXECUTING
        context = context or {}
        
        queries = await self.plan(task, context)
        all_findings = []
        
        for query in queries:
            results = await self.search(query)
            all_findings.extend(results)
        
        synthesis = await self.synthesize(all_findings)
        
        self.state = AgentState.COMPLETED
        return {
            "queries": queries,
            "findings": all_findings,
            "synthesis": synthesis
        }
    
    async def plan(self, task: str, context: Dict[str, Any]) -> List[str]:
        """Generate search queries for the research task."""
        return [task]
    
    async def execute(self, task: str, context: Optional[Dict[str, Any]] = None) -> AgentResult:
        """Execute a research task."""
        self.logger.info("research_execute", task=task[:100] if task else "")
        self.state = AgentState.EXECUTING
        
        try:
            result = await self.run(task, context)
            self.state = AgentState.COMPLETED
            return AgentResult(
                success=True,
                data=result,
                metadata={"queries_executed": len(result.get("queries", []))}
            )
        except Exception as e:
            self.state = AgentState.ERROR
            self.logger.error("research_error", error=str(e))
            return AgentResult(success=False, error=str(e))
    
    async def initialize(self) -> None:
        """Initialize the research agent."""
        await super().initialize()
        self._search_cache = {}
        self.logger.info("research_agent_initialized")
    
    async def shutdown(self) -> None:
        """Shutdown the research agent."""
        self._search_cache.clear()
        await super().shutdown()
        self.logger.info("research_agent_shutdown")
